src/lender/tranche/v3/tranche.sol

```diff
@@ -0,0 +1,169 @@
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program.  If not, see <https://www.gnu.org/licenses/>.
+
+pragma solidity >=0.5.15 <0.6.0;
+
+import "tinlake-auth/auth.sol";
+import "tinlake-math/math.sol";
+
+contract ERC20Like {
+  function balanceOf(address) public view returns (uint);
+  function transferFrom(address,address,uint) public returns (bool);
+  function mint(address, uint) public;
+  function burn(address, uint) public;
+  function totalSupply() public view returns (uint);
+}
+
+contract TickerLike {
+    function currentEpoch() public returns (uint);
+}
+
+contract Tranche is Math, Auth {
+mapping (uint => Epoch) public epochs;
+struct Epoch {
+    uint totalRedeem;
+    uint totalSupply;
+    // denominated in WAD
+    // percentage ONE == 100%
+    uint redeemFulfillment;
+    // denominated in WAD
+    // percentage ONE == 100%
+    uint supplyFulfillment;
+    // tokenPrice after end of epoch
+    uint tokenPrice; 
+    mapping (address => uint) supplyCurrencyAmount;
+    mapping (address => uint) redeemTokenAmount;
+}
+
+ERC20Like public currency;
+ERC20Like public token;
+TickerLike public ticker;
+address self;
+
+constructor(address currency_, address token_, address ticker_) public {
+    wards[msg.sender] = 1;
+    token = ERC20Like(token_);
+    currency = ERC20Like(currency_);
+    ticker = TickerLike(ticker_);
+    self = address(this);
+}
+
+function supplyCurrencyAmount(uint epochID, address addr) public view returns (uint) {
+    Epoch storage epoch = epochs[epochID];
+    return epoch.supplyCurrencyAmount[addr];
+}
+
+function redeemTokenAmount(uint epochID, address addr) public view returns (uint) {
+    Epoch storage epoch = epochs[epochID];
+    return epoch.redeemTokenAmount[addr];
+}
+
+function balance() external view returns (uint) {
+    return currency.balanceOf(self);
+}
+
+function tokenSupply() external view returns (uint) {
+    return token.totalSupply();
+}
+
+function depend(bytes32 contractName, address addr) public auth {
+    if (contractName == "token") { token = ERC20Like(addr); }
+    if (contractName == "currency") { currency = ERC20Like(addr); }
+    if (contractName == "ticker") { ticker = TickerLike(addr); }
+    else revert();
+}
+
+// supplyOrder function can be used to place or revoke an supply 
+function supplyOrder(uint epochID, uint supplyAmount) public {
+    require((epochID >= ticker.currentEpoch()), "epoch-already-over");
+    uint currentSupplyAmount = epochs[epochID].supplyCurrencyAmount[msg.sender];
+    epochs[epochID].supplyCurrencyAmount[msg.sender] = supplyAmount;
+    epochs[epochID].totalSupply = safeAdd(safeSub(epochs[epochID].totalSupply, currentSupplyAmount), supplyAmount);
+    if (supplyAmount > currentSupplyAmount) {
+        uint delta = safeSub(supplyAmount, currentSupplyAmount);
+        require(currency.transferFrom(msg.sender, self, delta), "currency-transfer-failed");
```

Comment:
why no `delta>0` check here?

Reply:
Because supplyAmount is bigger then currentSupplyAmount so delta <= 0 is not possible